src/common/settings_common.h

```diff
@@ -0,0 +1,254 @@
+// SPDX-FileCopyrightText: Copyright 2023 yuzu Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
+#include <forward_list>
+#include <functional>
+#include <map>
+#include <string>
+#include <typeindex>
+#include "common/common_types.h"
+
+namespace Settings {
+
+enum class Category : u32 {
+    Audio,
+    Core,
+    Cpu,
+    CpuDebug,
+    CpuUnsafe,
+    Renderer,
+    RendererAdvanced,
+    RendererDebug,
+    System,
+    SystemAudio,
+    DataStorage,
+    Debugging,
+    DebuggingGraphics,
+    Miscellaneous,
+    Network,
+    WebService,
+    AddOns,
+    Controls,
+    Ui,
+    UiGeneral,
+    UiLayout,
+    UiGameList,
+    Screenshots,
+    Shortcuts,
+    Multiplayer,
+    Services,
+    Paths,
+    MaxEnum,
+};
+
+constexpr u8 SpecializationTypeMask = 0xf;
+constexpr u8 SpecializationAttributeMask = 0xf0;
+constexpr u8 SpecializationAttributeOffset = 4;
+
+enum Specialization : u8 {
+    Default = 0,
+    Time = 1,
+    Hex = 2,
+    List = 3,
+    RuntimeList = 4,
+    Scalar = 5,
+    Countable = 6,
+    Paired = 7,
+
+    Percentage = (1 << SpecializationAttributeOffset),
+};
+
+bool IsConfiguringGlobal();
+void SetConfiguringGlobal(bool is_global);
+
+class BasicSetting;
+
+class Linkage {
+public:
+    explicit Linkage(u32 initial_count = 0);
+    ~Linkage();
+    std::map<Category, std::forward_list<BasicSetting*>> by_category{};
+    std::vector<std::function<void()>> restore_functions{};
+    u32 count;
+};
+
+/**
+ * BasicSetting is an abstract class that only keeps track of metadata. The string methods are
+ * available to get data values out.
+ */
+class BasicSetting {
+protected:
+    explicit BasicSetting(Linkage& linkage, const std::string& name, enum Category category_,
```

Comment:
```suggestion
    explicit BasicSetting(Linkage& linkage, const std::string& name, Category category_,
```
`enum class`es are typenames in C++, so I don't think you need to specifically say it's an enum when declaring a variable with it?